File: src/main/java/mediathek/update/ProgramUpdateCheck.java
package mediathek.update;

import mediathek.config.Daten;
import mediathek.config.Konstanten;
import mediathek.config.MVConfig;
import mediathek.daten.DatenPset;
import mediathek.daten.ListePset;
import mediathek.daten.ListePsetVorlagen;
import mediathek.gui.dialog.DialogNewSet;
import mediathek.mainwindow.MediathekGui;
import mediathek.tool.GuiFunktionenProgramme;
import mediathek.tool.NetUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.swing.*;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

/**
 * Perform check for updates every 24 hours if program is running long enough.
 */
public class ProgramUpdateCheck implements AutoCloseable {
    private static final Logger logger = LogManager.getLogger(ProgramUpdateCheck.class);
    private final Daten daten;
    private ScheduledFuture<?> actionFuture;

    public ProgramUpdateCheck(Daten daten) {
        this.daten = daten;
    }

    private void searchForProgramUpdate() {
        var pgrUpdate = new ProgrammUpdateSuchen();
        pgrUpdate.checkVersion(false, true, false, true);
    }

    private void checkForPsetUpdates() {
        try {
            SwingUtilities.invokeLater(() -> {
                final var parent = MediathekGui.ui();
                ListePset listePsetStandard = ListePsetVorlagen.getStandarset(parent, false);
                String version = MVConfig.get(MVConfig.Configs.SYSTEM_VERSION_PROGRAMMSET);
                if (listePsetStandard != null) {
                    if (!Daten.listePset.isEmpty()) {
                        // ansonsten ist die Liste leer und dann gibts immer was
                        if (listePsetStandard.version.isEmpty()) {
                            // dann hat das Laden der aktuellen Standardversion nicht geklappt
                            return;
                        }
                        if (version.equals(listePsetStandard.version)) {
                            // dann passt alles
                            return;
                        } else {
                            DialogNewSet dialogNewSet = new DialogNewSet(parent);
                            dialogNewSet.setVisible(true);
                            if (!dialogNewSet.ok) {
                                logger.info("Setanlegen: Abbruch");
                                if (!dialogNewSet.morgen) {
                                    // dann auch die Versionsnummer aktualisieren
                                    logger.info("Setanlegen: Nicht wieder nachfragen");
                                    MVConfig.add(MVConfig.Configs.SYSTEM_VERSION_PROGRAMMSET, listePsetStandard.version);
                                }
                                // dann halt nicht
                                return;
                            }
                        }
                    }

                    //========================================
                    // gibt keine Sets oder aktualisieren
                    // damit die Variablen ersetzt werden
                    ListePset.progMusterErsetzen(parent, listePsetStandard);

                    MVConfig.add(MVConfig.Configs.SYSTEM_VERSION_PROGRAMMSET, listePsetStandard.version);
                    // die Zielpafade anpassen
                    ListePset listePsetOrgSpeichern = Daten.listePset.getListeSpeichern();
                    if (!listePsetOrgSpeichern.isEmpty()) {
                        for (DatenPset psNew : listePsetStandard.getListeSpeichern()) {
                            psNew.arr[DatenPset.PROGRAMMSET_ZIEL_PFAD] = listePsetOrgSpeichern.get(0).arr[DatenPset.PROGRAMMSET_ZIEL_PFAD];
                            psNew.arr[DatenPset.PROGRAMMSET_THEMA_ANLEGEN] = listePsetOrgSpeichern.get(0).arr[DatenPset.PROGRAMMSET_THEMA_ANLEGEN];
                            psNew.arr[DatenPset.PROGRAMMSET_LAENGE_BESCHRAENKEN] = listePsetOrgSpeichern.get(0).arr[DatenPset.PROGRAMMSET_LAENGE_BESCHRAENKEN];
                            psNew.arr[DatenPset.PROGRAMMSET_LAENGE_FIELD_BESCHRAENKEN] = listePsetOrgSpeichern.get(0).arr[DatenPset.PROGRAMMSET_LAENGE_FIELD_BESCHRAENKEN];
                            psNew.arr[DatenPset.PROGRAMMSET_MAX_LAENGE] = listePsetOrgSpeichern.get(0).arr[DatenPset.PROGRAMMSET_MAX_LAENGE];
                            psNew.arr[DatenPset.PROGRAMMSET_MAX_LAENGE_FIELD] = listePsetOrgSpeichern.get(0).arr[DatenPset.PROGRAMMSET_MAX_LAENGE_FIELD];
                        }
                    }
                    if (!Daten.listePset.isEmpty()) {
                        // wenn leer, dann gibts immer die neuen und die sind dann auch aktiv
                        for (DatenPset psNew : listePsetStandard) {
                            // die bestehenden Sets sollen nicht gestört werden
                            psNew.arr[DatenPset.PROGRAMMSET_IST_ABSPIELEN] = Boolean.FALSE.toString();
                            psNew.arr[DatenPset.PROGRAMMSET_IST_ABO] = Boolean.FALSE.toString();
                            psNew.arr[DatenPset.PROGRAMMSET_IST_BUTTON] = Boolean.FALSE.toString();
                            psNew.arr[DatenPset.PROGRAMMSET_IST_SPEICHERN] = Boolean.FALSE.toString();
                        }
                        // damit man sie auch findet :)
                        String date = new SimpleDateFormat("dd.MM.yyyy").format(new Date());
                        listePsetStandard.forEach((psNew) -> psNew.arr[DatenPset.PROGRAMMSET_NAME] = psNew.arr[DatenPset.PROGRAMMSET_NAME] + ", neu: " + date);
                    }
                    GuiFunktionenProgramme.addSetVorlagen(MediathekGui.ui(), daten, listePsetStandard, true); // damit auch AddOns geladen werden
                    logger.info("Setanlegen: OK");
                    logger.info("==========================================");
                }
            });
        } catch (Exception ignored) {
        }
    }

    private void performUpdateCheck() {
        logger.debug("performUpdateCheck started.");
        var gui = MediathekGui.ui();
        try {
            //first check if network is available...
            if (NetUtils.isReachable("res.mediathekview.de",1000)) {
                //we have internet...
                SwingUtilities.invokeLater(() -> gui.enableUpdateMenuItem(false));

                var externalUpdateCheck = System.getProperty(Konstanten.EXTERNAL_UPDATE_PROPERTY);
                if (externalUpdateCheck == null || !externalUpdateCheck.equalsIgnoreCase("true")) {
                    searchForProgramUpdate();
                }
                else {
                    logger.info("External Update Mechanism in use -> skip program update check");
                }

                checkForPsetUpdates();
            } else
                logger.warn("Update Check: Network is not reachable.");
        }
        finally {
            SwingUtilities.invokeLater(() -> gui.enableUpdateMenuItem(true));
        }
        logger.debug("performUpdateCheck finished.");
    }

    public void start() {
        logger.debug("ProgramUpdateCheck Started.");
        actionFuture = daten.getTimerPool().scheduleWithFixedDelay(() -> SwingUtilities.invokeLater(this::performUpdateCheck), 60L, TimeUnit.SECONDS.convert(24L, TimeUnit.HOURS), TimeUnit.SECONDS);
    }

    @Override
    public void close() {
        if (actionFuture != null) {
            actionFuture.cancel(true);
        }
        logger.debug("ProgramUpdateCheck closed.");
    }
}

File: src/main/kotlin/mediathek/update/ProgramUpdateCheck.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.update

import kotlinx.coroutines.*
import kotlinx.coroutines.swing.Swing
import mediathek.config.application.ApplicationConfiguration
import mediathek.daten.*
import mediathek.gui.dialog.DialogNewSet
import mediathek.tool.GuiFunktionen
import mediathek.tool.GuiFunktionenProgramme
import mediathek.tool.NetUtils
import org.apache.logging.log4j.LogManager
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.function.BiConsumer
import javax.swing.JFrame
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.seconds

/**
 * Perform check for updates every 24 hours if program is running long enough.
 */
class ProgramUpdateCheck(
    private val host: ProgramUpdateHost,
    private val programSets: ProgramSetRepository,
    private val programSetExporter: BiConsumer<Array<DatenPset>, String>,
) : AutoCloseable {
    private val job = SupervisorJob()
    private val scope = CoroutineScope(job + Dispatchers.IO + CoroutineExceptionHandler { _, ex ->
        logger.error("Program update check failed", ex)
    })
    private val programUpdateSearch = ProgrammUpdateSuchen(host::ownerFrame, scope)
    private var updateCheckJob: Job? = null

    fun start() {
        logger.debug("ProgramUpdateCheck Started.")
        updateCheckJob?.cancel()
        updateCheckJob = scope.launch {
            delay(INITIAL_UPDATE_CHECK_DELAY)
            while (isActive) {
                performUpdateCheck()
                delay(UPDATE_CHECK_INTERVAL)
            }
        }
    }

    private suspend fun performUpdateCheck() {
        logger.debug("performUpdateCheck started.")
        var updateMenuItemDisabled = false
        try {
            if (NetUtils.isReachable(UPDATE_CHECK_HOST, 1.seconds)) {
                withContext(Dispatchers.Swing) {
                    host.enableUpdateMenuItem(false)
                }
                updateMenuItemDisabled = true

                if (GuiFunktionen.isNotUsingExternalUpdater()) {
                    searchForProgramUpdate()
                } else {
                    logger.info("External Update Mechanism in use -> skip program update check")
                }

                checkForPsetUpdates()
            } else {
                logger.warn("Update Check: Network is not reachable.")
            }
        } finally {
            if (updateMenuItemDisabled) {
                withContext(NonCancellable + Dispatchers.Swing) {
                    host.enableUpdateMenuItem(true)
                }
            }
            logger.debug("performUpdateCheck finished.")
        }
    }

    private suspend fun searchForProgramUpdate() {
        programUpdateSearch.checkVersionSuspending(
            showAlert = false,
            showProgramInformation = true,
            showAllInformation = false,
            silent = true,
            showErrors = false,
        )
    }

    private suspend fun checkForPsetUpdates() {
        val standardPset = withContext(Dispatchers.IO) {
            ListePsetVorlagen.getStandarset(null, false)
        } ?: return

        withContext(Dispatchers.Swing) {
            checkForPsetUpdatesOnSwingThread(standardPset)
        }
    }

    private fun checkForPsetUpdatesOnSwingThread(standardPset: ListePset) {
        val parent = host.ownerFrame()
        if (!shouldInstallStandardPset(parent, standardPset)) {
            return
        }

        installStandardPset(parent, standardPset)
    }

    private fun shouldInstallStandardPset(parent: JFrame, standardPset: ListePset): Boolean {
        if (currentPsets().isEmpty()) {
            return true
        }

        if (standardPset.version.isEmpty()) {
            return false
        }

        val installedVersion = ApplicationConfiguration.getInstance().standardProgramSetVersion
        if (installedVersion == standardPset.version) {
            return false
        }

        return confirmStandardPsetUpdate(parent, standardPset)
    }

    private fun confirmStandardPsetUpdate(parent: JFrame, standardPset: ListePset): Boolean {
        val dialogNewSet = DialogNewSet(parent, programSets)
        dialogNewSet.isVisible = true
        val decision = dialogNewSet.decision
        if (decision.accepted) {
            return true
        }

        logger.info("Setanlegen: Abbruch")
        if (!decision.askAgainTomorrow) {
            logger.info("Setanlegen: Nicht wieder nachfragen")
            updateInstalledStandardPsetVersion(standardPset)
        }
        return false
    }

    private fun installStandardPset(parent: JFrame, standardPset: ListePset) {
        ProgramSetTemplateResolver.replaceTemplates(parent, standardPset)

        updateInstalledStandardPsetVersion(standardPset)
        copySaveSettingsFromExistingSet(standardPset)
        prepareImportedSetsForExistingConfiguration(standardPset)

        GuiFunktionenProgramme.addSetVorlagen(parent, programSets, standardPset, true, programSetExporter)
        logger.info("Setanlegen: OK")
        logger.info("==========================================")
    }

    private fun updateInstalledStandardPsetVersion(standardPset: ListePset) {
        ApplicationConfiguration.getInstance().standardProgramSetVersion = standardPset.version
    }

    private fun copySaveSettingsFromExistingSet(standardPset: ListePset) {
        val existingSaveSets = currentPsets().listeSpeichern
        if (existingSaveSets.isEmpty()) {
            return
        }

        val existingPset = existingSaveSets.first()
        for (newPset in standardPset.listeSpeichern) {
            newPset.zielPfad = existingPset.zielPfad
            newPset.isThemaAnlegen = existingPset.isThemaAnlegen
            newPset.isLaengeBeschraenken = existingPset.isLaengeBeschraenken
            newPset.isLaengeFieldBeschraenken = existingPset.isLaengeFieldBeschraenken
            newPset.maxLaenge = existingPset.maxLaenge
            newPset.maxLaengeField = existingPset.maxLaengeField
        }
    }

    private fun prepareImportedSetsForExistingConfiguration(standardPset: ListePset) {
        if (currentPsets().isEmpty()) {
            return
        }

        disableImportedSets(standardPset)
        markImportedSetsAsNew(standardPset)
    }

    private fun disableImportedSets(standardPset: ListePset) {
        for (newPset in standardPset) {
            newPset[DatenPset.PROGRAMMSET_IST_ABSPIELEN] = false.toString()
            newPset[DatenPset.PROGRAMMSET_IST_ABO] = false.toString()
            newPset[DatenPset.PROGRAMMSET_IST_BUTTON] = false.toString()
            newPset[DatenPset.PROGRAMMSET_IST_SPEICHERN] = false.toString()
        }
    }

    private fun markImportedSetsAsNew(standardPset: ListePset) {
        val date = LocalDate.now().format(NEW_SET_DATE_FORMATTER)
        standardPset.forEach { newPset ->
            newPset.name += NEW_SET_NAME_SUFFIX_PREFIX + date
        }
    }

    private fun currentPsets(): ListePset = programSets.list

    override fun close() {
        job.cancel()
        logger.debug("ProgramUpdateCheck closed.")
    }

    private companion object {
        private val logger = LogManager.getLogger(ProgramUpdateCheck::class.java)
        private const val UPDATE_CHECK_HOST = "res.mediathekview.de"
        private val INITIAL_UPDATE_CHECK_DELAY = 60.seconds
        private val UPDATE_CHECK_INTERVAL = 24.hours
        private const val NEW_SET_NAME_SUFFIX_PREFIX = ", neu: "
        private val NEW_SET_DATE_FORMATTER = DateTimeFormatter.ofPattern("dd.MM.yyyy")
    }
}
